Catalog: Categorize Items, Vertical Tabs

dtaylor113 opened the pull request:
![catalog](https://user-images.githubusercontent.com/12733153/46016765-a679c300-c0a3-11e8-9ca4-0e4606a4a1ca.gif)

https://jira.coreos.com/browse/CONSOLE-757
https://jira.coreos.com/browse/CONSOLE-846

dtaylor113 commented:
Still working on restricting vertical tabs to just show the tree/tabs for the active item.

spadgett commented on frontend/public/components/utils/categorize-catalog-items.tsx:
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable no-undef */
```
Shouldn't be a ts file if there are no components.

But this should move into the existing catalog component file since it's only used there.

dtaylor113 replied:
These are a set of methods dealing with categorization.  catalog-items.jsx is going to have a lot of code for filtering, vert. tabs, sorting, etc.. Trying to make code more modular

spadgett commented on frontend/public/components/utils/categorize-catalog-items.tsx:
```diff
@@ -0,0 +1,109 @@
+
+const hasMatchingTags = (tagsOne, tagsTwo) => {
+  return _.some(tagsOne, (tagOne) => {
+    let tagOneLower: string = tagOne.toLowerCase();
```
const

You should declare tagOne and tagTwo as strings in the signature. Then the type is inferred here

spadgett commented on frontend/public/components/utils/categorize-catalog-items.tsx:
```diff
@@ -0,0 +1,109 @@
+    subCategory.items = _.isArray(subCategory.items) ? subCategory.items.concat([item]) : [item];
+  }
+  category.items = _.isArray(category.items) && !_.find(category.items, ['id', item.id]) ? category.items.concat([item]) : [item];
+  return true;
```
Why do we have a return value here if it's always true

dtaylor113 replied:
For `itemCategorized` flag in outer loop/calling method

spadgett replied:
I guess I don't understand... It always returns true. How does checking the return value help?

dtaylor113 replied:
It determines if an item was categorized or if not should go into 'Other' category.

spadgett commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+
+  renderTabs(category, parentCategoryId=null){
+    if (!category.numItems) {
+      return;
```
return null;

spadgett commented on frontend/public/components/utils/categorize-catalog-items.tsx:
```diff
@@ -0,0 +1,109 @@
+const catalogCategories: any = [
+  {id: 'all', label: 'All Categories'},
+  {id: 'languages', label: 'Languages', subCategories: [
+    {id: 'java', label: 'Java', tags: ['java'], icon: 'icon-openjdk'},
```
Let's remove icons since they're not used

spadgett commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+      onActivate={() => this.onActivateTab(parentCategoryId ? [parentCategoryId, category.id] : [category.id])}
+      hasActiveDescendant={this.doesHaveActiveDescendants(category.id)}
+      shown={category.id === 'all'}>
+      {category.subCategories && category.numItems && <VerticalTabs key={`${category.id}-tabs`}
```
category.numItems is always true because we check it above

I don't think you need the `-tabs` on then end of key

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,203 @@
+    };
+
+    if (props.items) {
+      this.allCategories = categorizeItems(props.items);
```
You can declare state as a class item and then avoid the constructor all together:

```
export class CatalogTileViewPage extends React.Component {
  state = {
    showAllItemsForCategory: null,
    activeTabs: ['all'], // array of tabs [main category, sub-category]
  };
```

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+
+  renderCategoryTabs() {
+    return (
+      <VerticalTabs restrictTabs={true} activeTab={this.isAllTabActive()} shown="true">
```
don't need shown="true" here

jeff-phillips-18 replied:
activeTab should be true if any of the top level categories or All, is active

jeff-phillips-18 replied:
can replace `restrictTabs={true} ` with simply `restrictTabs`

dtaylor113 replied:
> activeTab should be true if any of the top level categories or All, is active

I believe when I had that, all top level tabs displayed even when drilled down into sub-categories.  I believe we only want to show the active 'tree' of the active tab, not other parent/main categories.

spadgett replied:
Do we really want to hide the other top level categories? That seems strange to me. There aren't that many.

jeff-phillips-18 replied:
That is the design, @tlwu2013

jeff-phillips-18 replied:
That is the design, @tlwu2013?

dtaylor113 replied:
@tlwu2013, @serenamarie125, we do need to decide on this.  IMO it does seem a little jarring to hide top level categories when drilling down to sub-cats, plus it requires and extra click on 'All Cats' to bring back the top level categories.

spadgett replied:
It also moves the position of the category you just clicked, which is a little annoying if you wanted to click a subcategory. They don't expand below your mouse cursor where you'd expect.

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+
+  doesHaveActiveDescendants(categoryId) {
+    const { activeTabs } = this.state;
+    return activeTabs[0] === categoryId;
```
This is fine if limiting to 2 levels. You could check if any of the activeTabs have the categoryId to support more levels. Better to keep just one active tab Id, and determine if a parent has an active descendant based on that.

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+      hasActiveDescendant={this.doesHaveActiveDescendants(category.id)}
+      shown={category.id === 'all'}>
+      {category.subCategories && category.numItems && <VerticalTabs key={`${category.id}-tabs`}
+        restrictTabs={true} activeTab={this.activeTabIsSubCategory(category.subCategories)}>
```
can replace `restrictTabs={true} ` with simply `restrictTabs`

dtaylor113 commented:
> If it's this difficult to add the categories, we might want to revisit how we're doing it. We don't need to be backwards compatible, so we can do something simpler if needed.

I'm open to suggestions :-) It's pretty straight-forward to me, probably 'cause this is second time implementing it :-)  We have an obj. with hard coded Categories and Sub-Categories.  We add 'items', based on item tags, to arrays under either a category or sub-category.  This allows us to know the hierarchy and also do 'numItem' counts for each category/sub-category.

It is actually a little more confusing if we/I add back in the code which dealt with custom categories where an item could match a main custom Category, but if didn't match a custom Sub-Category, then item would go into an 'Other' sub-cat for the custom Category.

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
+      <div className="row">
+        <div className="col-sm-2">
+          <FilterSidePanel id="filter-panel">
+            { this.renderCategoryTabs() }
```
This should not be inside a `FilterSidePanel` as that causes it to have a 15px padding and not align correctly.

jeff-phillips-18 commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,204 @@
-  items: PropTypes.array.isRequired
+    return (
+      <div className="row">
+        <div className="col-sm-2">
```
I don't believe we want a bootstrap grid here. The eventual use of the filterSidePanel will set the width of the sidebar. The outer div should be `display: flex` the main content should have `flex: 1`.

dtaylor113 commented:
Hi, I believe I have addressed all review comments.  Updated video in description.  Thanks @jeff-phillips-18 & @spadgett for the code review comments.  @spadgett, I will be studying the PR you pushed for days, it is a good example of 'clean code' and advanced lodash -thanks :-)

spadgett commented on frontend/public/components/_catalog-item.scss:
```diff
@@ -66,3 +66,26 @@ a.co-catalog-item-tile {
   }
 }
+
+.co-catalog-page {
```
We should be following BEM naming here:

* co-catalog-page
* co-catalog-page__tabs
* co-catalog-page__content
* co-catalog-page__heading
* co-catalog-page__num-items

spadgett commented on frontend/public/components/catalog-items.jsx:
```diff
@@ -1,48 +1,201 @@
-  </CatalogTileView>;
+  render() {
+    const { activeTabs, showAllItemsForCategory, currentCategories, numItems } = this.state;
+    const activeCategory = showAllItemsForCategory ? _.find(currentCategories, { id: showAllItemsForCategory }) : null;
```
I'm not sure the `_.find` check is right here... `state.showAllItemsForCategory` is the category object itself, right?

(This might be my bug...)

dtaylor113 replied:
no `showAllItemsForCategory` is the id of a category :-)

tlwu2013 commented:
Spot couple of styling differences from this PR comparing to redline mock shared previously. I'll post the mocks with more information for reference.

tlwu2013 commented:
Hi @dtaylor113 , in general it looks good, thanks!

Here are the redlines for Catalog View. I've added more spacing/color info on them:

**Res: 1280px**
![catalog-view__redline_1280px](https://user-images.githubusercontent.com/5903705/46043878-084f2280-c0ce-11e8-9d2d-259b0ed99be5.png)

The idea here is using the "panel" to wrap the Category tabs and the tiles to emphasize they are correlated (so that the Category tabs are not visually competing with Console's side-nav on the left). 

For responsiveness, here are the responsive redlines from mobile to Full HD:
**Responsive:**
![catalog__view__responsive_redline](https://user-images.githubusercontent.com/5903705/46044090-ac38ce00-c0ce-11e8-9e3b-b15077790a5c.png)

Also, the breadcrumbs (missing in this PR) is for highlighting the category hierarchy:
![catalog-view__l3_redline_1280px](https://user-images.githubusercontent.com/5903705/46044353-66303a00-c0cf-11e8-9421-ce9677a7f7c0.png)

Here are the "category tab"/"view all" interaction redlines:
![catalog__view-all__flow_redlines](https://user-images.githubusercontent.com/5903705/46044476-b5766a80-c0cf-11e8-8337-21390d678683.png)

Please let me know if any questions, thanks!

spadgett commented:
Thanks @tlwu2013. I suggest we ask @rhamilto to look at the styles in a follow up PR.

spadgett commented:
Created a sub-task CONSOLE-872 for updating the styles.

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

serenamarie125 commented:
@tlwu2013 we do not want a breadcrumb in these mocks, can you pull them please ?

@spadgett let's make sure that the breadcrumb is not included in the implementation.  thanks!

spadgett commented:
Created sub-task CONSOLE-873 to add the breadcrumb.

spadgett commented:
Sorry I misread... We should *not* include the breadcrumb, correct?

spadgett commented:
jenkins rebuild

serenamarie125 commented:
correct @spadgett do NOT include the breadcrumb, thanks!

rhamilto commented:
@tlwu2013, the vertical spacing between tiles is currently 20px per the upstream component.  Are you good with that or do we need to change it upstream?
